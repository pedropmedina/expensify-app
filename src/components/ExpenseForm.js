import React from 'react';

class ExpenseForm extends React.Component {
	state = { fields: { description: '', amount: '', note: '' } };

	onFieldChange = event => {
		const name = event.target.name;
		const val = event.target.value;
		const fields = this.state.fields;

		if (name === 'amount') {
			if (val.match(/^\d*(\.\d{0,2})?$/)) {
				fields[name] = val;
			}
		} else fields[name] = val;

		this.setState({ fields });
	};

	render() {
		const { description, amount, note } = this.state.fields;
		return (
			<form>
				<input
					type="text"
					name="description"
					value={description}
					placeholder="Description"
					autoFocus
					onChange={this.onFieldChange}
				/>
				<input
					type="text"
					name="amount"
					value={amount}
					placeholder="Amount"
					onChange={this.onFieldChange}
				/>
				<textarea
					name="note"
					value={note}
					placeholder="Add a note for  the expense"
					onChange={this.onFieldChange}
				/>
				<button>add expense</button>
			</form>
		);
	}
}

export default ExpenseForm;
