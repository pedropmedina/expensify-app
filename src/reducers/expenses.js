export const ADD_EXPENSE = 'ADD_EXPENSE';
export const EDIT_EXPENSE = 'EDIT_EXPENSE';
export const REMOVE_EXPENSE = 'REMOVE_EXPENSE';

const expensesReducerDefaultState = [];
const expenseReducer = (state = expensesReducerDefaultState, action) => {
	switch (action) {
		case ADD_EXPENSE:
			return [...state, action.expense];
		case EDIT_EXPENSE:
			return state.map(expense => {
				if (expense.id === action.id) {
					return { ...expense, ...action.updates };
				} else {
					return expense;
				}
			});
		case REMOVE_EXPENSE:
			return state.filter(({ id }) => id !== action.id);
		default:
			return state;
	}
};

export default expenseReducer;
