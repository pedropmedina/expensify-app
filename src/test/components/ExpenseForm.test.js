import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';

import ExpenseForm from '../../components/ExpenseForm';

import expenses from '../fixtures/expenses';

test('should render ExpenseForm correclty', () => {
	const wrapper = shallow(<ExpenseForm />);
	expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseForm with data', () => {
	const wrapper = shallow(<ExpenseForm expense={expenses[1]} />);
	expect(wrapper).toMatchSnapshot();
});

test('should redner error for invaldid form submission', () => {
	const wrapper = shallow(<ExpenseForm />);
	expect(wrapper).toMatchSnapshot();
	wrapper.find('form').simulate('submit', { preventDefault: () => {} });
	expect(wrapper.state('error').length).toBeGreaterThan(0);
	expect(wrapper).toMatchSnapshot();
});

test('should set description on input change', () => {
	const value = 'New description';
	const name = 'description';
	const wrapper = shallow(<ExpenseForm />);
	wrapper
		.find('input')
		.at(0)
		.simulate('change', { target: { value, name } });
	expect(wrapper.state().fields[name]).toBe(value);
});

test('should set note on textarea change', () => {
	const value = 'New note value';
	const name = 'note';
	const wrapper = shallow(<ExpenseForm />);
	wrapper.find('textarea').simulate('change', { target: { name, value } });
	expect(wrapper.state().fields[name]).toBe(value);
});

test('should set amount id valid input', () => {
	const name = 'amount';
	const value = '23.50';
	const wrapper = shallow(<ExpenseForm />);
	wrapper
		.find('input')
		.at(1)
		.simulate('change', { target: { name, value } });
	expect(wrapper.state().fields[name]).toBe(value);
});

test('should not set amount if invalid input', () => {
	const name = 'amount';
	const value = '12.122';
	const wrapper = shallow(<ExpenseForm />);
	wrapper
		.find('input')
		.at(1)
		.simulate('change', { target: { name, value } });
	expect(wrapper.state().fields[name]).toBe('');
});

test('should call onSubmit prop for valid form submission', () => {
	const { description, amount, note, createdAt } = expenses[0];
	const onSubmitSpy = jest.fn();
	const wrapper = shallow(
		<ExpenseForm {...expenses[0]} onSubmit={onSubmitSpy} />,
	);
	wrapper.find('form').simulate('submit', { preventDefault: () => {} });
	expect(wrapper.state('error')).toBe('');
	expect(onSubmitSpy).toHaveBeenLastCalledWith({
		description,
		amount,
		note,
		createdAt,
	});
});
