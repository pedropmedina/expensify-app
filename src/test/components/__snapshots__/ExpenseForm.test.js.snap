// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should redner error for invaldid form submission 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      name="description"
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      name="amount"
      onChange={[Function]}
      placeholder="Amount"
      type="text"
      value=""
    />
    <DatePicker
      allowSameDay={false}
      dateFormat="L"
      dateFormatCalendar="MMMM YYYY"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      fixedHeight={true}
      isClearable={true}
      monthsShown={1}
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      placeholderText="Select date for expense"
      preventOpenOnFocus={false}
      selected={"1970-01-01T00:00:00.000Z"}
      shouldCloseOnSelect={true}
      showTimeSelect={false}
      timeCaption="Time"
      timeIntervals={30}
      todayButton="Today"
      withPortal={false}
    />
    <textarea
      name="note"
      onChange={[Function]}
      placeholder="Add a note for the expense (optional)"
      value=""
    />
    <button>
      add expense
    </button>
  </form>
</div>
`;

exports[`should redner error for invaldid form submission 2`] = `
<div>
  <p>
    Please provide description and amount
  </p>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      name="description"
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      name="amount"
      onChange={[Function]}
      placeholder="Amount"
      type="text"
      value=""
    />
    <DatePicker
      allowSameDay={false}
      dateFormat="L"
      dateFormatCalendar="MMMM YYYY"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      fixedHeight={true}
      isClearable={true}
      monthsShown={1}
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      placeholderText="Select date for expense"
      preventOpenOnFocus={false}
      selected={"1970-01-01T00:00:00.000Z"}
      shouldCloseOnSelect={true}
      showTimeSelect={false}
      timeCaption="Time"
      timeIntervals={30}
      todayButton="Today"
      withPortal={false}
    />
    <textarea
      name="note"
      onChange={[Function]}
      placeholder="Add a note for the expense (optional)"
      value=""
    />
    <button>
      add expense
    </button>
  </form>
</div>
`;

exports[`should render ExpenseForm correclty 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      name="description"
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      name="amount"
      onChange={[Function]}
      placeholder="Amount"
      type="text"
      value=""
    />
    <DatePicker
      allowSameDay={false}
      dateFormat="L"
      dateFormatCalendar="MMMM YYYY"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      fixedHeight={true}
      isClearable={true}
      monthsShown={1}
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      placeholderText="Select date for expense"
      preventOpenOnFocus={false}
      selected={"1970-01-01T00:00:00.000Z"}
      shouldCloseOnSelect={true}
      showTimeSelect={false}
      timeCaption="Time"
      timeIntervals={30}
      todayButton="Today"
      withPortal={false}
    />
    <textarea
      name="note"
      onChange={[Function]}
      placeholder="Add a note for the expense (optional)"
      value=""
    />
    <button>
      add expense
    </button>
  </form>
</div>
`;

exports[`should render ExpenseForm with data 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      name="description"
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      name="amount"
      onChange={[Function]}
      placeholder="Amount"
      type="text"
      value=""
    />
    <DatePicker
      allowSameDay={false}
      dateFormat="L"
      dateFormatCalendar="MMMM YYYY"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      fixedHeight={true}
      isClearable={true}
      monthsShown={1}
      onBlur={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      placeholderText="Select date for expense"
      preventOpenOnFocus={false}
      selected={"1970-01-01T00:00:00.000Z"}
      shouldCloseOnSelect={true}
      showTimeSelect={false}
      timeCaption="Time"
      timeIntervals={30}
      todayButton="Today"
      withPortal={false}
    />
    <textarea
      name="note"
      onChange={[Function]}
      placeholder="Add a note for the expense (optional)"
      value=""
    />
    <button>
      add expense
    </button>
  </form>
</div>
`;
